compiler/modules/front/codeGen.bal

```diff
@@ -716,11 +739,55 @@ function codeGenIfElseStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environ
                 elseContBlock.insns.push(branch);
                 assignments.push(...elseAssignments);
             }
-            return { block: contBlock, assignments };
+            StmtNarrowing[] narrowings = exprStmtNarrowing(exprNarrowing, ifContBlock != (), elseContBlock != ());
+            // pr-todo: should this be also pushUnion?
+            narrowings.push(...ifNarrowings);
+            foreach var n in elseNarrowings {
+                pushUnion(narrowings, n);
+            }
+            return { block: contBlock, assignments, narrowings };
+        }
+    }
+}
+
+// Convert if-else expression narrowing to a statement narrowing
+function exprStmtNarrowing(Narrowing? exprNarrowing, boolean ifCompletes, boolean elseCompletes) returns StmtNarrowing[] {
+    StmtNarrowing[] ns = [];
+    if exprNarrowing != () {
+        if ifCompletes {
+            // pr-todo: recheck result
+            bir:InsnResult ifResult = { insn: exprNarrowing.testInsn, result: !exprNarrowing.negated };
+            StmtNarrowing ifNarrowing = { binding: exprNarrowing.binding, basis: ifResult, completeNormally: exprNarrowing.ifTrue };
+            ns.push(ifNarrowing);
+        }
+        if elseCompletes {
+            bir:InsnResult elseResult = { insn: exprNarrowing.testInsn, result: exprNarrowing.negated };
+            StmtNarrowing elseNarrowing = { binding: exprNarrowing.binding, basis: elseResult, completeNormally: exprNarrowing.ifFalse };
+            // pr-todo: this should be pushUnion
+            ns.push(elseNarrowing);
+        }
+    }
+    return ns;
+}
+
+
+// if already exists in the list, union, else append
+function pushUnion(StmtNarrowing[] list, StmtNarrowing annex) {
```

Comment:
I would call this something like `addStmtNarrowing`.

Reply:
I have a counterpart to this now, so picked a different name.